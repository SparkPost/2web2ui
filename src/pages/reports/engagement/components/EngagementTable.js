import React from 'react';
import _ from 'lodash';
import styled from 'styled-components';
import { InfoOutline } from '@sparkpost/matchbox-icons';
import { Box, Panel, Tooltip } from 'src/components/matchbox';
import useUniqueId from 'src/hooks/useUniqueId';
import { Empty, PanelLoading, TableCollection } from 'src/components';
import { map as METRICS_MAP } from 'src/config/metrics';
import { formatNumber, formatPercent } from 'src/helpers/units';
import { safeRate } from 'src/helpers/math';

const StyledTooltipIcon = styled.div`
  margin-left: 5px;
  transform: translateY(-1px);
`;

const COLUMNS = [
  {
    key: 'link_name',
    label: 'Link',
    sortKey: 'link_name',
    width: '30%',
  },
  {
    key: 'count_raw_clicked_approx',
    label: (
      <HeaderTooltip tooltipContent={METRICS_MAP.count_raw_clicked_approx.description}>
        Unique Clicks
      </HeaderTooltip>
    ),
    sortKey: 'count_raw_clicked_approx',
    width: '30%',
  },
  {
    key: 'count_clicked',
    label: (
      <HeaderTooltip tooltipContent={METRICS_MAP.count_clicked.description}>Clicks</HeaderTooltip>
    ),
    sortKey: 'count_clicked',
    width: '20%',
  },
  {
    key: 'percentage_clicked',
    label: 'Percent of Total',
    sortKey: 'percentage_clicked',
    width: '20%',
  },
];

const DataRow = row => [
  row.link_name,
  formatNumber(row.count_raw_clicked_approx),
  formatNumber(row.count_clicked),
  formatPercent(row.percentage_clicked),
];

export default function EngagementTable({ data, loading }) {
  if (loading) {
    return <PanelLoading />;
  }

  if (data.length === 0) {
    return (
      <Panel.LEGACY>
        <Empty message="No clicks to report" />
      </Panel.LEGACY>
    );
  }

  // Manually count the total number of clicks (FWIW didn't included metric with chart data request,
  // and use the result here to avoid the dependency)
  const totalClicks = _.sumBy(data, 'count_clicked');

  // Must include percentage in data for sorting
  const dataWithPercentage = data.map(row => ({
    ...row,
    percentage_clicked: safeRate(row.count_clicked, totalClicks),
  }));

  return (
    <TableCollection
      columns={COLUMNS}
      defaultSortColumn="count_clicked"
      defaultSortDirection="desc"
      getRowData={DataRow}
      pagination
      rows={dataWithPercentage}
    />
  );
}

function HeaderTooltip({ children, tooltipContent }) {
  const id = useUniqueId('header-tooltip');

  return (
    <span>
      {children}
      <Tooltip id={id} dark top content={tooltipContent}>
        <Box as="span" color="blue.700">
          <StyledTooltipIcon as={InfoOutline} />
        </Box>
      </Tooltip>
    </span>
  );
}
