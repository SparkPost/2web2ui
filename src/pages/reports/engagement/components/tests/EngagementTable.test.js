import React from 'react';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

import EngagementTable from '../EngagementTable';

cases('EngagementTable', (props) => {
  const wrapper = shallow(<EngagementTable {...props} />);
  expect(wrapper).toMatchSnapshot();
}, {
  'when loading': { loading: true },
  'when no results': { data: [], loading: false },
  'when renders table': {
    data: [
      { count_raw_clicked_approx: 1, count_clicked: 10, link_name: 'Raw URL' },
      { count_raw_clicked_approx: 2, count_clicked: 20, link_name: 'Nintendo' }
    ],
    loading: false
  }
});
