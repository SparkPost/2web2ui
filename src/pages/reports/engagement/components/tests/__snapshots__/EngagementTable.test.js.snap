// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngagementTable when loading 1`] = `
<PanelLoading
  accent={false}
  minHeight="400px"
/>
`;

exports[`EngagementTable when no results 1`] = `
<Panel.LEGACY>
  <Empty
    message="No clicks to report"
  />
</Panel.LEGACY>
`;

exports[`EngagementTable when renders table 1`] = `
<TableCollection
  columns={
    Array [
      Object {
        "key": "link_name",
        "label": "Link",
        "sortKey": "link_name",
        "width": "30%",
      },
      Object {
        "key": "count_raw_clicked_approx",
        "label": <HeaderTooltip
          tooltipContent="Total number of messages which had at least one link selected one or more times."
        >
          Unique Clicks
        </HeaderTooltip>,
        "sortKey": "count_raw_clicked_approx",
        "width": "30%",
      },
      Object {
        "key": "count_clicked",
        "label": <HeaderTooltip
          tooltipContent="Number of times that links were clicked across all emails."
        >
          Clicks
        </HeaderTooltip>,
        "sortKey": "count_clicked",
        "width": "20%",
      },
      Object {
        "key": "percentage_clicked",
        "label": "Percent of Total",
        "sortKey": "percentage_clicked",
        "width": "20%",
      },
    ]
  }
  defaultSortColumn="count_clicked"
  defaultSortDirection="desc"
  getRowData={[Function]}
  pagination={true}
  rows={
    Array [
      Object {
        "count_clicked": 10,
        "count_raw_clicked_approx": 1,
        "link_name": "Raw URL",
        "percentage_clicked": 33.33333333333333,
      },
      Object {
        "count_clicked": 20,
        "count_raw_clicked_approx": 2,
        "link_name": "Nintendo",
        "percentage_clicked": 66.66666666666666,
      },
    ]
  }
/>
`;
